File: src/main/java/com/tim/sernachat/service/JwtTokenProvider.java
package com.tim.sernachat.service;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.Jws;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.MalformedJwtException;
import io.jsonwebtoken.SignatureException;
import io.jsonwebtoken.UnsupportedJwtException;
import io.jsonwebtoken.security.Keys;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Date;
import javax.crypto.SecretKey;


@Slf4j
@Component
public class JwtTokenProvider {
    @Value("${spring.jwt.secret}")
    private String secretKey;

    private long tokenValid = 1000L * 60 * 60;

    @SuppressWarnings("deprecated")
    public String generateToken(String name) {

        SecretKey key = Keys.hmacShaKeyFor(secretKey.getBytes(StandardCharsets.UTF_8));

        Date now = Date.from(Instant.now());
        return Jwts.builder()
                .setId(name)
                .setIssuedAt(now)
                .setExpiration(new Date(now.getTime() + tokenValid))
                .signWith(key)
                .compact();
    }

    public String getUserFromToken(String token) {
        return getClaims(token).getBody().getId();
    }

    public boolean validateToken(String token) {
        return this.getClaims(token) != null;
    }

    @SuppressWarnings("deprecated")
    private Jws<Claims> getClaims(String token) {

        try {
            return Jwts.parserBuilder().setSigningKey(Keys.hmacShaKeyFor(secretKey.getBytes(StandardCharsets.UTF_8))).build().parseClaimsJws(token);
        } catch (SignatureException ex) {
            log.error("Invalid JWT signature");
            throw ex;
        } catch (MalformedJwtException ex) {
            log.error("Invalid JWT token");
            throw ex;
        } catch (ExpiredJwtException ex) {
            log.error("Expired JWT token");
            throw ex;
        } catch (UnsupportedJwtException ex) {
            log.error("Unsupported JWT token");
            throw ex;
        } catch (IllegalArgumentException ex) {
            log.error("JWT claims string is empty");
            throw ex;
        }
    }

}

File: src/main/kotlin/com/itime/sernachat/service/JwtTokenProvider.kt
package com.itime.sernachat.service

import io.jsonwebtoken.*
import io.jsonwebtoken.security.Keys
import org.slf4j.Logger
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.nio.charset.StandardCharsets
import java.time.Instant
import java.util.*

@Component
class JwtTokenProvider(val logger: Logger) {
    @Value("\${spring.jwt.secret}")
    private val secretKey: String? = null
    private val tokenValid = 1000L * 60 * 60
    fun generateToken(name: String?): String {
        val key = Keys.hmacShaKeyFor(secretKey!!.toByteArray(StandardCharsets.UTF_8))
        val now = Date.from(Instant.now())
        return Jwts.builder()
            .setId(name)
            .setIssuedAt(now)
            .setExpiration(Date(now.time + tokenValid))
            .signWith(key)
            .compact()
    }

    fun getUserFromToken(token: String): String {
        return getClaims(token)!!.body.id
    }

    fun validateToken(token: String): Boolean {
        return getClaims(token) != null
    }

    private fun getClaims(token: String): Jws<Claims>? {
        return try {
            Jwts.parserBuilder().setSigningKey(Keys.hmacShaKeyFor(secretKey!!.toByteArray(StandardCharsets.UTF_8)))
                .build().parseClaimsJws(token)
        } catch (ex: SignatureException) {
            logger.error("Invalid JWT signature")
            throw ex
        } catch (ex: MalformedJwtException) {
            logger.error("Invalid JWT token")
            throw ex
        } catch (ex: ExpiredJwtException) {
            logger.error("Expired JWT token")
            throw ex
        } catch (ex: UnsupportedJwtException) {
            logger.error("Unsupported JWT token")
            throw ex
        } catch (ex: IllegalArgumentException) {
            logger.error("JWT claims string is empty")
            throw ex
        }
    }
}
